pyanaconda/installation.py

```diff
@@ -53,355 +53,377 @@ def _writeKS(ksdata):
-        late_storage = TaskQueue("Late storage configuration", N_("Configuring storage"))
-        conf_task = storage_proxy.WriteConfigurationWithTask()
-        late_storage.append_dbus_tasks(STORAGE, [conf_task])
-        installation_queue.append(late_storage)
-
-    # Do bootloader.
-    bootloader_proxy = STORAGE.get_proxy(BOOTLOADER)
-    bootloader_install = TaskQueue("Bootloader installation", N_("Installing boot loader"))
-
-    def run_configure_bootloader():
-        tasks = boss_proxy.CollectConfigureBootloaderTasks(
-            payload.kernel_version_list
+        write_configs = TaskQueue(
+            "Write configs and kickstarts",
+            N_("Storing configuration files and kickstarts")
         )
 
-        for service, task in tasks:
-            sync_run_task(DBus.get_proxy(service, task))
-
-    bootloader_install.append(Task("Configure bootloader", run_configure_bootloader))
-
-    def run_install_bootloader():
-        tasks = bootloader_proxy.InstallBootloaderWithTasks(
-            payload.type,
-            payload.kernel_version_list
+        # Write the kickstart file to the installed system (or, copy the input
+        # kickstart file over if one exists).
+        if conf.target.can_save_output_kickstart:
+            # write anaconda related configs & kickstarts
+            write_configs.append(Task("Store kickstarts", _writeKS, (ksdata,)))
+        else:
+            # don't write the kickstart file to the installed system if this has
+            # been disabled by the nosave option
+            log.warning("Writing of the output kickstart to installed system has been disabled"
+                        " by the nosave option.")
+
+        # only add write_configs to the main queue if we actually store some kickstarts/configs
+        if write_configs.task_count:
+            configuration_queue.append(write_configs)
+
+        post_scripts = TaskQueue(
+            "Post installation scripts",
+            N_("Running post-installation scripts")
+        )
+        post_scripts.append(Task(
+            "Run post installation scripts",
+            runPostScripts,
+            (ksdata.scripts,)
+        ))
+        configuration_queue.append(post_scripts)
+
+        boss_proxy = BOSS.get_proxy()
+        finish_tasks = boss_proxy.FinishInstallationWithTasks()
+        configuration_queue.append_dbus_tasks(BOSS, finish_tasks)
+
+        return configuration_queue
+
+    def _wait_for_threads_to_finish(self):
+        """Wait for background processing threads to finish.
+
+        Wait for background processing threads to finish before filling
+        the installation task queue. Otherwise installation tasks might
+        be created based on old data in DBus modules, missing data set by
+        the threads that are still running.
+        """
+
+        # This should be the only thread running, wait for the others to finish if not.
+        if threadMgr.running > 1:
```

Comment:
I guess that this code have to change in the future. But not something for this PR. Although, even now it will not work for Storage thread.

Reply:
Definitely. However, the Storage thread should be fine. We are still using UI threads for everything. It is one the areas that we will have to look into very soon. The plan for this task is to run it via DBus from the Web UI (with some limitations), but  GUI and TUI will still run it directly (to avoid the limitations). There is too much work to be done.